### PR TITLE
Fix CommandInvocationParser when missing closing parenthesis

### PR DESCRIPTION
**Pull request by FlorianChevassu**

### CODE REVIEW

**Review by Xaxetrov — COMMENTED** [review]
Seems good (did not test myself).
This parsing function begins to be quite long. I think it would be good to either split it into smaller explicit functions or to comment the parts.

**Comment by FlorianChevassu** [comment]
Agreed. I will update the PR with a bit of refactoring.

File: src/CommandInvocationParser.cpp [hunk 1]
```diff
@@ -18,6 +18,75 @@ namespace cmake::language
     {
       return { std::find_if_not(r.begin, r.end, [](char c) {return std::isspace(c); }), r.end };
     }
+
+    //-----------------------------------------------------------------------------
+    tl::expected<Token, Error> ParserIdentifier(Range r)
```
Comment (Xaxetrov) [inline src/CommandInvocationParser.cpp]: `ParserIdentifier` -> `ParseIdentifier` ?

If so, also change `ParserArguments`

File: src/CommandInvocationParser.cpp [hunk 2]
```diff
@@ -18,6 +18,75 @@ namespace cmake::language
...
+      auto newRange = IgnoreSpaces(r);
+      if (*newRange.begin != '(')
+      {
+        Error err;
+        err.message = "A command invocation should wrap its arguments between parenthesis.";
+        err.context = newRange.begin;
+        return tl::make_unexpected(err);
+      }
+
+      auto args = Parser<ElementType::Arguments>{}.Parse(Range{ newRange.begin + 1, r.end });
```
Comment (Xaxetrov) [inline src/CommandInvocationParser.cpp:47]: General reflection maybe not directly for the PR here: I find it strange to remove the first parenthesis from what we pass to the arguments parser but to keep the end parenthesis.

File: src/CommandInvocationParser.cpp [hunk 3]
```diff
@@ -18,6 +18,75 @@ namespace cmake::language
...
+          Error err;
+          err.message = "Missing parenthesis";
+          err.context = args->range.end;
+          return tl::make_unexpected(err);
+        }
+
+        if (*args->range.end != ')')
+        {
+          Error err;
+          err.message = fmt::format("Expected \')\', got {}", *args->range.end);
```
Comment (Xaxetrov) [inline src/CommandInvocationParser.cpp]: We already checked that the opening parenthesis is `(` so the error should be "got `}`" IMO.

File: src/CommandInvocationParser.cpp [hunk 4]
```diff
@@ -18,6 +18,75 @@ namespace cmake::language
...
+          err.context = newRange.begin + 1;
+          return tl::make_unexpected(err);
+        }
+        if (*(newRange.begin + 1) != ')')
+        {
+          Error err;
+          err.message = fmt::format("Expected \')\', got {}", *(newRange.begin + 1));
+          err.context = newRange.begin + 1;
+          return tl::make_unexpected(err);
+        }
```
Comment (Xaxetrov) [inline src/CommandInvocationParser.cpp]: For the last two error cases, do you see a way to factorize them with the precedent ones which are very similar, while keeping the readability to an ok level ?

File: src/CommandInvocationParser.cpp (hunk 2 above)
Comment (FlorianChevassu) [reply src/CommandInvocationParser.cpp:47]: Here we do not know where the arguments ends, we only know `newRange.end+1` is the beginning of the arguments. `r.end` points to the end of the range, which usually points to the end of the script, not just the end of the command.

File: src/CommandInvocationParser.cpp (hunk 3 above)
Comment (FlorianChevassu) [reply src/CommandInvocationParser.cpp]: Here, the `{}` are just placeholders that will be replaced by the value of `*args->range.end`.

File: src/CommandInvocationParser.cpp (hunk 4 above)
Comment (FlorianChevassu) [reply src/CommandInvocationParser.cpp]: Yes, I think we should be able to construct a range in the `if/else` and then check for an error / return a result. I will look at it.

**Review by Xaxetrov — APPROVED** [review]

**Pull request merged**